### ISSUE TITLE
[critical] Uniform diff check should update only known uniforms

### ISSUE THREAD

**Issue by eXponenta** [pmndrs/react-ogl#8]
Bug in this place:
https://github.com/pmndrs/react-ogl/blob/7d9800111cc20d37578a05a094a8f2d6c17f7ab2/src/utils.ts#L95

when you use a special material with internal state and custom uniforms, react will produce errors because default uniforms will be dropped.

This is critical, because witout this will not working a special components for program with attach!

for ex:

```ts

import { Program } from "ogl";
import { extend, Node } from "react-ogl";

import React from "react";

const FRAG = `
precision highp float;

uniform vec3 uColor;
uniform vec2 uPoint;
uniform sampler2D uSampler0;

varying vec3 vNormal;
varying vec2 vUv;

void main() {
	gl_FragColor = texture2D(uSampler0, vUv);
	gl_FragColor = mix (vec4(1.0, 0., 0., 1.), gl_FragColor, step(0.01, length(uPoint - vUv)));
}
`;

const VERT = `
attribute vec3 position;
attribute vec3 normal;
attribute vec2 uv;

uniform mat4 modelViewMatrix;
uniform mat4 projectionMatrix;
uniform mat3 normalMatrix;

varying vec3 vNormal;
varying vec2 vUv;

void main() {
	vNormal = normalize(normalMatrix * normal);
	vUv = uv;
	gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
}
`;

export interface IProgProps {
	mousePoint?: Array<number>;
	color?: string | any;
	children: any[];
}

export class BaseProgramWithTexture extends Program {
	constructor(gl: GLContext) {
		super(gl, {
			vertex: VERT,
			fragment: FRAG,
			transparent: true,
			uniforms:{
				uColor: { value: [1,1,1] },
				uPoint: { value: [0,0] },
				uSampler0: { value: null },
			}
		});
	}

	set texture(v) {
		this.uniforms.uSampler0 = {value: v};
	}

	get texture() {
		return this.uniforms.uSampler0?.value;
	}
}

extend({BaseProgramWithTexture});

declare global {
	namespace JSX {
	  interface IntrinsicElements {
		baseProgramWithTexture: Node<BaseProgramWithTexture, typeof BaseProgramWithTexture>
	  }
	}
  }


export default React.forwardRef<Program, IProgProps>(
	({ color = "pink", mousePoint = [0, 0], children }, ref) => {
		return (
			<baseProgramWithTexture
				ref={ref}
				uniforms = {{
					uColor: color
				}}
			>
                        <ChessTexture/>
			</baseProgramWithTexture>
		);
	}
);

// ChessTexture.tsx
// which should apply texture to `texture` field that is setter
export default ({width = 256, height = 256, step = 64}) => {
	return React.createElement('texture', {
		attach: 'texture',
		image: generateCheckmate(width, height, step)
	})
}


//programRef is reference on to our baseProgramWithTexture and uPoint MUST BE AVAILABLE
	const handleMove = ({ hit }: any) => {
		(
			programRef.current.uniforms as Record<string, { value: any }>
		).uPoint.value = hit?.uv || [0, 0];
	};
```

Will have a a lot of warns:
![image](https://user-images.githubusercontent.com/12572222/158614447-bfa968f5-845a-4169-89f3-65b910f4d022.png)

And if we try to change uPoint manually - will crash, because reconciler remove required fields.

**Comment by eXponenta** [comment]
Possible solution for:
https://github.com/pmndrs/react-ogl/blob/7d9800111cc20d37578a05a094a8f2d6c17f7ab2/src/utils.ts#L95

https://github.com/eXponenta/react-ogl/pull/1/commits/9df24f3383e877f1679d8480a7cd1934ffef25cb

and remove reduce to prefer forEach or `for in` because we will apply properties immediately

**Comment by CodyJasonBennett** [comment]
This doesn't look good. Maybe send in a PR, we can figure this out and add coverage for this stuff.

**Comment by CodyJasonBennett** [comment]
Thanks. Fix published in [`v0.1.1`](https://github.com/pmndrs/react-ogl/releases/tag/v0.1.1) alongside #7 (hiked up a patch release since 0.1.0 was locked).
